fdbserver/workloads/BitFlipping.actor.cpp

```diff
@@ -28,85 +28,202 @@
-        }
-        return Void();
-    }
-
-    ACTOR template <class W>
-    Future<Void> flipBitsClient(Database cx, BitFlippingWorkload* self) {
-        state double lastTime = now();
-        state double workloadEnd = now() + self->testDuration;
-        state std::vector<W> machines;
-        loop {
-            wait(poisson(&lastTime, 1));
-            wait(BitFlippingWorkload::getAllStorageWorkers(cx, self, &machines));
-            auto machine = deterministicRandom()->randomChoice(machines);
-            self->doBitFlips(machine, self->percentBitFlips);
-        }
-    }
+	bool enabled;
+	double testDuration;
+	double percentBitFlips;
+	double periodicCheckInterval;
+	std::vector<NetworkAddress> chosenWorkers;
+	std::vector<Future<Void>> clients;
+
+	BitFlippingWorkload(WorkloadContext const& wcx) : TestWorkload(wcx) {
+		enabled = !clientId; // only do this on the "first" client
+		testDuration = getOption(options, LiteralStringRef("testDuration"), 10.0);
+		percentBitFlips = getOption(options, LiteralStringRef("percentBitFlips"), 10.0);
+		periodicCheckInterval = getOption(options, LiteralStringRef("periodicCheckInterval"), 10.0);
+	}
+
+	std::string description() const override {
+		if (&g_simulator == g_network)
+			return "BitFlipping";
+		else
+			return "NoSimBitFlipping";
+	}
+
+	Future<Void> setup(Database const& cx) override { return Void(); }
+
+	// Starts the workload by -
+	// 1. Starting the actor to periodically check chaosMetrics, and
+	// 2. Starting the actor that injects failures on chosen storage servers
+	Future<Void> start(Database const& cx) override {
+		if (enabled) {
+			clients.push_back(periodicMetricCheck(this));
+			clients.push_back(flipBitsClient<WorkerInterface>(cx, this));
+			return timeout(waitForAll(clients), testDuration, Void());
+		} else
+			return Void();
+	}
+
+	Future<bool> check(Database const& cx) override { return true; }
+
+	void getMetrics(vector<PerfMetric>& m) override {}
+
+	static void checkBitFlipResult(Future<Void> res, WorkerInterface worker) {
+		if (res.isError()) {
+			auto err = res.getError();
+			if (err.code() == error_code_client_invalid_operation) {
+				TraceEvent(SevError, "ChaosDisabled")
+				    .detail("OnEndpoint", worker.waitFailure.getEndpoint().addresses.address.toString());
+			} else {
+				TraceEvent(SevError, "BitFlippingFailed")
+				    .detail("OnEndpoint", worker.waitFailure.getEndpoint().addresses.address.toString())
+				    .error(err);
+			}
+		}
+	}
+
+	ACTOR void doBitFlips(WorkerInterface worker, double percentage, double startDelay = 0.0) {
+		state Future<Void> res;
+		wait(::delay(startDelay));
+		SetFailureInjection::FlipBitsCommand flipBits;
+		flipBits.percentBitFlips = percentage;
+		SetFailureInjection req;
+		req.flipBits = flipBits;
+		res = worker.clientInterface.setFailureInjection.getReply(req);
+		wait(ready(res));
+		checkBitFlipResult(res, worker);
+	}
+
+	ACTOR static Future<Void> getAllStorageWorkers(Database cx,
```

Comment:
This function is not necessary.
```
wait(getAllStorageWorkers(a, b, c));
```
is equivalent to
```
wait(store(c, getStorageWorkers(a, b));
```
or without using store
```
std::vector<WorkerInterface> v = wait(getStorageWorkers(a, b));
c = std::move(v);
```

Reply:
Ok